synapse/handlers/presence.py

```diff
@@ -365,10 +365,9 @@ def _handle_timeouts(self):
                     if now - last_update > EXTERNAL_PROCESS_EXPIRY
                 ]
                 for process_id in expired_process_ids:
-                    users_to_check.update(
+                    users_to_check.update(self.wheel_timer.fetch(
                         self.external_process_last_updated_ms.pop(process_id, ())
-                    )
-                    self.external_process_last_update.pop(process_id)
+                    ))
```

Comment:
I don't think this is quite what we want to be doing, as:

- The `external_process_last_updated_ms` is a mapping from the external processes to the last time we saw an update, which will always be in the past
- `wheel_timer.fetch` gets all entries that have timed out *before* the given timestamp, i.e. usually the timestamp given is "now". Given that the timestamp passed into it is in the past, it will return a subset of users that were fetched already when we did `wheel_time.fetch(now)` above.

What we're trying to achieve is to find all users that were syncing on processes that have died and then remove that process from the `external_process_*` dicts. So maybe something like the following?

```python
for process_id in expired_process_ids:
    # Get all the users that were syncing on the dead process, and remove the process
    users_on_process = self.external_process_to_current_syncs.pop(process_id, ())

    users_to_check.update(users_on_process)  # We need to check those users

    # Remove the dead process from the dicts
    self.external_process_last_updated_ms.pop(process_id, ())
```

Reply:
Would you like to create the commit yourself or can I pick your code and update my commit?

Reply:
Feel free to commit it yourself if you agree with it :)